File: src/main/scala/de/mineformers/allomancy/world/MetalGenerator.scala
package de.mineformers.allomancy.world

import java.util.Random

import de.mineformers.allomancy.block.AllomanticMetalOre
import net.minecraft.util.BlockPos
import net.minecraft.world.World
import net.minecraft.world.chunk.IChunkProvider
import net.minecraft.world.gen.feature.WorldGenMinable
import net.minecraftforge.fml.common.IWorldGenerator

/**
  * MetalGenerator
  *
  * @author PaleoCrafter
  */
class MetalGenerator extends IWorldGenerator {

  case class Ore(range: Range, veins: Int, veinSize: Int)

  final val Properties = Map(
    "copper" -> Ore(0 to 64, 20, 9),
    "zinc" -> Ore(0 to 64, 20, 9),
    "tin" -> Ore(0 to 64, 20, 9),
    "aluminium" -> Ore(0 to 64, 20, 9),
    "chromium" -> Ore(0 to 64, 20, 9)
  )
  final val Generators = Properties.map(e =>
    e._1 -> new WorldGenMinable(AllomanticMetalOre.fromMetal(e._1), e._2.veinSize))

  override def generate(random: Random,
                        chunkX: Int,
                        chunkZ: Int,
                        world: World,
                        chunkGenerator:
                        IChunkProvider,
                        chunkProvider: IChunkProvider): Unit =
    if (world.provider.getDimensionId == 0) {
      val x = chunkX * 16
      val z = chunkZ * 16
      val iterator = Properties.iterator
      while (iterator.hasNext) {
        val entry = iterator.next()
        var i = 0
        while (i < entry._2.veins) {
          Generators(entry._1)
            .generate(world, random, new BlockPos(
              x + random.nextInt(16),
              random.nextInt(entry._2.range.max - 1) + entry._2.range.min,
              z + random.nextInt(16)))
          i += 1
        }
      }
    }
}

File: src/main/java/de/mineformers/investiture/allomancy/world/MetalGenerator.java
package de.mineformers.investiture.allomancy.world;

import com.google.common.collect.ImmutableMap;
import de.mineformers.investiture.allomancy.Allomancy;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraft.world.chunk.IChunkProvider;
import net.minecraft.world.gen.IChunkGenerator;
import net.minecraft.world.gen.feature.WorldGenMinable;
import net.minecraftforge.fml.common.IWorldGenerator;

import java.util.Map;
import java.util.Random;
import java.util.stream.Collectors;

/**
 * Generates ores for all allomantic metals.
 */
public class MetalGenerator implements IWorldGenerator
{
    /**
     * Container class for each ore's configuration
     */
    private static class Ore
    {
        final int minY;
        final int maxY;
        final int veins;
        final int veinSize;

        Ore(int minY, int maxY, int veins, int veinSize)
        {
            this.minY = minY;
            this.maxY = maxY;
            this.veins = veins;
            this.veinSize = veinSize;
        }
    }

    // Should be replaced by real config
    private final static Map<String, Ore> ORES = ImmutableMap
        .of("copper", new Ore(0, 64, 20, 9),
            "zinc", new Ore(0, 64, 20, 9),
            "tin", new Ore(0, 64, 20, 9),
            "aluminium", new Ore(0, 64, 20, 9),
            "chromium", new Ore(0, 64, 20, 9));
    private final static Map<String, WorldGenMinable> GENERATORS = ORES.entrySet().stream().collect(
        Collectors.toMap(Map.Entry::getKey, e -> new WorldGenMinable(Allomancy.Blocks.ALLOMANTIC_ORE.fromMetal(e.getKey()), e.getValue().veinSize)));

    @Override
    public void generate(Random random, int chunkX, int chunkZ, World world, IChunkGenerator chunkGenerator, IChunkProvider chunkProvider)
    {
        if (world.provider.getDimension() == 0)
        {
            int x = chunkX * 16;
            int z = chunkZ * 16;
            for (Map.Entry<String, Ore> entry : ORES.entrySet())
            {
                String metal = entry.getKey();
                Ore ore = entry.getValue();
                for (int i = 0; i < ore.veins; i++)
                {
                    GENERATORS.get(metal).generate(world, random,
                                                   new BlockPos(x + random.nextInt(16),
                                                                random.nextInt(ore.maxY - ore.minY + 1) + ore.minY,
                                                                z + random.nextInt(16)));
                }
            }
        }
    }
}
